### ISSUE TITLE
Task 2: Implement LSTM and GRU networks for sequential modeling

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ts-rs#48]
# Task 2: Implement LSTM and GRU Networks for Sequential Modeling

**Related to Epic**: #46  
**Depends on**: #47

## Objective
Implement Long Short-Term Memory (LSTM) and Gated Recurrent Unit (GRU) neural networks optimized for time series forecasting and sequence modeling.

## Scope
- Create `src/ml/recurrent.rs` with LSTM and GRU implementations
- Implement sequence-to-sequence and sequence-to-one architectures
- Add bidirectional and multi-layer support
- Integrate with time series forecasting pipeline

### Core Functions to Implement
```rust
// Main recurrent network interface
pub fn create_lstm_forecaster(
    config: &LSTMConfig,
    training_data: &TimeSeries,
) -> Result<LSTMForecaster, MLError>;

pub fn create_gru_forecaster(
    config: &GRUConfig,
    training_data: &TimeSeries,
) -> Result<GRUForecaster, MLError>;

// Training and prediction
pub fn train_recurrent_model(
    model: &mut dyn RecurrentModel,
    data: &TimeSeriesDataset,
    config: &TrainingConfig,
) -> Result<TrainingResult, MLError>;

pub fn forecast_with_lstm(
    model: &LSTMForecaster,
    input_sequence: &[f64],
    horizon: usize,
) -> Result<ForecastResult, MLError>;
```

### Data Structures
```rust
pub struct LSTMForecaster {
    pub network: NeuralNetwork,
    pub sequence_length: usize,
    pub features: Vec<String>,
    pub normalization: NormalizationParams,
    pub training_stats: TrainingStats,
}

pub struct LSTMConfig {
    pub hidden_size: usize,
    pub num_layers: usize,
    pub sequence_length: usize,
    pub dropout_rate: f32,
    pub bidirectional: bool,
    pub output_size: usize,
    pub feature_engineering: FeatureConfig,
}

pub struct GRUConfig {
    pub hidden_size: usize,
    pub num_layers: usize,
    pub sequence_length: usize,
    pub dropout_rate: f32,
    pub bidirectional: bool,
    pub output_size: usize,
    pub reset_gate_bias: f32,
}

pub struct TimeSeriesDataset {
    pub sequences: Vec<Vec<f64>>,
    pub targets: Vec<Vec<f64>>,
    pub timestamps: Vec<DateTime<Utc>>,
    pub features: Vec<String>,
    pub split_info: DataSplit,
}
```

### 1. LSTM Architecture
- **Cell State Management**: Forget gate, input gate, output gate implementation
- **Multi-layer Support**: Stacked LSTM layers with residual connections
- **Bidirectional Processing**: Forward and backward sequence processing
- **Attention Mechanism**: Optional attention layer for long sequences

### 2. GRU Architecture
- **Simplified Gating**: Reset gate and update gate mechanisms
- **Computational Efficiency**: Faster training compared to LSTM
- **Multi-layer Support**: Stacked GRU layers
- **Gradient Flow**: Improved gradient flow for long sequences

### 3. Sequence Processing
- **Sequence-to-One**: Many timesteps to single prediction
- **Sequence-to-Sequence**: Input sequence to output sequence
- **Variable Length**: Support for variable-length sequences
- **Padding and Masking**: Handle sequences of different lengths

### 4. Time Series Integration
```rust
// Integration with existing forecasting
impl ForecastMethod for LSTMForecaster {
    fn forecast(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        // LSTM-specific forecasting logic
    }
    
    fn evaluate(&self, test_data: &TimeSeries) -> Result<ModelEvaluation, MLError> {
        // Performance evaluation metrics
    }
}

// Feature engineering for time series
pub struct FeatureConfig {
    pub include_lags: Vec<usize>,
    pub include_rolling_stats: bool,
    pub rolling_windows: Vec<usize>,
    pub include_seasonal_features: bool,
    pub seasonal_periods: Vec<usize>,
    pub include_trend_features: bool,
}
```

### 5. Training Optimizations
- **Gradient Clipping**: Prevent exploding gradients
- **Learning Rate Scheduling**: Adaptive learning rate adjustment
- **Early Stopping**: Prevent overfitting with validation monitoring
- **Batch Processing**: Efficient mini-batch training

### 1. Attention Mechanisms
```rust
pub struct AttentionLSTM {
    pub lstm_layers: Vec<LSTMLayer>,
    pub attention_layer: AttentionLayer,
    pub output_projection: DenseLayer,
}

// Attention for focusing on relevant timesteps
pub struct AttentionLayer {
    pub attention_weights: Tensor,
    pub context_vector: Tensor,
    pub alignment_model: AlignmentModel,
}
```

### 2. Regularization Techniques
- **Dropout**: Layer-wise dropout for regularization
- **Weight Decay**: L2 regularization on network weights
- **Batch Normalization**: Normalize layer inputs
- **Recurrent Dropout**: Dropout on recurrent connections

### 3. Model Interpretability
- **Attention Visualization**: Show which timesteps the model focuses on
- **Feature Importance**: Analyze which features are most important
- **Prediction Confidence**: Uncertainty quantification for predictions

### Forecasting Integration
- Extend `ForecastMethod` enum to include LSTM and GRU options
- Integrate with existing forecast evaluation framework
- Support confidence intervals and uncertainty quantification

### Data Preprocessing
- Use existing preprocessing pipeline for normalization
- Integrate with feature engineering capabilities
- Support missing value handling from existing modules

## Acceptance Criteria
- [ ] LSTM and GRU models train successfully on time series data
- [ ] Models achieve competitive forecasting accuracy vs existing methods
- [ ] Support for both univariate and multivariate time series
- [ ] GPU acceleration works correctly for training and inference
- [ ] Proper gradient flow without vanishing/exploding gradients
- [ ] Integration with existing forecasting evaluation framework
- [ ] Comprehensive test suite including edge cases
- [ ] Performance: Training on 10K datapoints completes in <10 minutes on GPU

## Test Cases Required
- Simple sine wave forecasting (should achieve high accuracy)
- Stock price prediction with multiple features
- Seasonal data with strong patterns
- Noisy data with weak signal
- Variable-length sequences
- Multi-step ahead forecasting

## Dependencies
- Task 1 results for core ML infrastructure
- `candle-core` and `candle-nn` for neural network implementation
- Existing `forecasting` module for integration
- `src/preprocessing` for data preparation

## Estimated Time
18-22 hours

## Next Steps
After completion, enables:
- Task 3: Transformer architecture implementation
- Task 4: Ensemble methods combining LSTM/GRU with other models
- Task 5: AutoML for automatic hyperparameter tuning
